File: frontend/src/types/production/analytics.ts
/** 单工序周期统计 */
export interface StepCycleStat {
  node_id: string
  node_name: string
  stage_name: string
  sort_order: number
  /** 样本数 */
  n: number
  /** 平均耗时（小时） */
  avg_hours: number
  /** 最短耗时（小时） */
  min_hours: number | null
  /** 最长耗时（小时） */
  max_hours: number | null
}

/** 工序周期分析响应 */
export interface StepCycleResponse {
  steps: StepCycleStat[]
  total_batches: number
  sample_note: string | null
}

/** 字段趋势点（跨批次时间序列） */
export interface FieldTrendPoint {
  batch_no: string
  filled_at: string
  value: number
}

/** 工段汇总平铺矩阵列定义（工序字段或计算字段） */
export interface StageSummaryColumn {
  /** 工序节点 id，前端按节点分组表头 */
  node_id: string
  node_code: string
  node_name: string
  field_key: string
  field_label: string
  unit: string | null
  kind: 'field' | 'computed'
  /** 行列扁平字典的键：{node_id}.{field_key}（node_code 仅路线内唯一，多路线会撞键） */
  col_key: string
}

/** 工段汇总平铺矩阵行：单批次一行，values/computed 键为 {node_id}.{field_key} */
export interface StageSummaryRow {
  batch_id: string
  batch_no: string
  values: Record<string, number | string | boolean | null>
  computed: Record<string, number | null>
}

/** 工段汇总平铺矩阵响应 */
export interface StageSummary {
  columns: StageSummaryColumn[]
  rows: StageSummaryRow[]
}

File: backend/app/modules/production/schemas/analytics.py
"""分析 API 契约。"""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


class StepCycleStat(BaseModel):
    """单工序周期统计"""
    model_config = ConfigDict(from_attributes=True)

    node_id: uuid.UUID
    node_name: str
    stage_name: str
    sort_order: int
    n: int
    avg_hours: float
    min_hours: float | None
    max_hours: float | None


class StepCycleResponse(BaseModel):
    """工序周期分析响应"""
    steps: list[StepCycleStat]
    total_batches: int
    sample_note: str | None = None


class FieldTrendPoint(BaseModel):
    """字段趋势数据点（跨批次时间序列）。"""

    batch_no: str
    filled_at: datetime
    value: float


class StageSummaryColumn(BaseModel):
    """工段汇总平铺矩阵列定义（工序字段或计算字段）。"""

    node_id: uuid.UUID  # 工序节点 id，前端按节点分组表头
    node_code: str
    node_name: str
    field_key: str
    field_label: str
    unit: str | None
    kind: Literal["field", "computed"]
    col_key: str  # 行列扁平字典的键：{node_id}.{field_key}（node_code 仅路线内唯一，多路线会撞键）


class StageSummaryRow(BaseModel):
    """工段汇总平铺矩阵行：单批次一行，values/computed 键为 {node_id}.{field_key}。"""

    batch_id: uuid.UUID
    batch_no: str
    values: dict[str, float | str | bool | None]
    computed: dict[str, float | None]


class StageSummaryOut(BaseModel):
    """工段汇总平铺矩阵响应。"""

    columns: list[StageSummaryColumn]
    rows: list[StageSummaryRow]
